### ISSUE TITLE
Crash while adding a duplicate entry of a song to the playlist

### ISSUE THREAD

**Issue by BumbleFlash** [CUTR-at-USF/MUSER#104]
**Describe the bug**

The app crashes in Pixel 3a when a duplicate entry of a song is added to the playlist. 

    The exception could not be delivered to the consumer because it has already canceled/disposed the flow or the exception has 
    nowhere to go, to begin with. Further reading: https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error- 
    handling | Bad window token, you cannot show a dialog before an Activity is created or after it's hidden.
            at io.reactivex.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:367)
            at io.reactivex.internal.observers.ConsumerSingleObserver.onSuccess(ConsumerSingleObserver.java:65)
            at io.reactivex.internal.operators.single.SingleObserveOn$ObserveOnSingleObserver.run(SingleObserveOn.java:81)
            at io.reactivex.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:119)
            at android.os.Handler.handleCallback(Handler.java:938)
            at android.os.Handler.dispatchMessage(Handler.java:99)
            at android.os.Looper.loop(Looper.java:223)
            at android.app.ActivityThread.main(ActivityThread.java:7656)
            at java.lang.reflect.Method.invoke(Native Method)
            at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
       Caused by: com.afollestad.materialdialogs.MaterialDialog$DialogException: Bad window token, you cannot show a dialog 
      before an Activity is created or after it's hidden.
            at com.afollestad.materialdialogs.MaterialDialog.show(MaterialDialog.java:466)
            at com.afollestad.materialdialogs.MaterialDialog$Builder.show(MaterialDialog.java:2189)
            at com.simplecity.amp_library.utils.playlists.PlaylistManager$addToPlaylist$1.accept(PlaylistManager.kt:131)
            at com.simplecity.amp_library.utils.playlists.PlaylistManager$addToPlaylist$1.accept(PlaylistManager.kt:39)

Currently, I'm testing it on other devices with API level < 29

**To Reproduce**
1. Add a song to a playlist.
2. Add the same song once again

**Expected behavior**
A dialog should pop up saying that the song already exists. 

**Observed behavior**
App crashes 

**Device and Android version:** 
Google Pixel 3a. Stock Android 11. Google Play Services 20.39.15

**Comment by BumbleFlash** [comment]
Looks like there is a `// fixme` added here indicating the `applicationContext` is the one causing the issue. 

https://github.com/CUTR-at-USF/MUSER/blob/track-favorites/app/src/main/java/com/simplecity/amp_library/utils/playlists/PlaylistManager.kt#L92

**Comment by BumbleFlash** [comment]
Update: The app crashes, per testing this issue on Moto G4+  Android 7.0,

**Comment by barbeau** [comment]
Yes, dialogs should be shown using the activity context. Here's a check that I've used in a utility method in OneBusAway and other apps for checking the context before displaying the dialog to be sure:

```java
    /**
     * Returns true if the activity is still active and dialogs can be managed (i.e., displayed
     * or dismissed), or false if it is
     * not
     *
     * @param activity Activity to check for displaying/dismissing a dialog
     * @return true if the activity is still active and dialogs can be managed, or false if it is
     * not
     */
    public static boolean canManageDialog(Activity activity) {
        if (activity == null) {
            return false;
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            return !activity.isFinishing() && !activity.isDestroyed();
        } else {
            return !activity.isFinishing();
        }
    }
```

Can you pass the activity context into this class in addition to the application context so you can use it for the dialog?

I would still use the application context for everything else (e.g., getting strings from resources, etc.).

**Comment by BumbleFlash** [comment]
@barbeau Sounds good. We can fix it on a separate PR once we merge #103?

**Comment by BumbleFlash** [comment]
Looking at the code, to the best of understanding, Shutte uses Dagger to inject the application context, and adding a new parameter in the constructor only throws a bunch of errors from the auto-generated `DaggerAppComponent`. Should we create a function instead in the `MainActivity` that returns the activity context and then use the above check to handle null cases?

**Comment by BumbleFlash** [comment]
Tentatively, I tried creating a static function that returns the static activity context of the `MainActivity` and then calling the function back in the `PlaylistManager.kt` and looks like it fixes the crash. Can this be a good alternative to what you suggested since `MainActivity` is the one where all the fragments in the view pager are attached to?

**Comment by barbeau** [comment]
@BumbleFlash Using static functions to get Activity contexts isn't a good idea because there can potentially be multiple instances of activities at the same time and you could get some strange behavior, including references to the wrong activity instance. I haven't used Dagger much myself so I'm not going to be much help without looking at it more closely. Can you look at injecting the activity context using Dagger?

**Comment by BumbleFlash** [comment]
@barbeau Sounds good. I haven't used Dagger either, I'll look into the documentation to understand it better and find a solution. 

Also, looks like `MainActivity` has a single instance launch mode in the Manifest. Does that mean only one instance would exist at the same run time, right?

**Comment by barbeau** [comment]
>Also, looks like MainActivity has a single instance launch mode in the Manifest. Does that mean only one instance would exist at the same run time, right?

Yes, that's correct:
https://developer.android.com/guide/topics/manifest/activity-element#lmode

I still wouldn't depend on this within the app for something like creating a static accessor to the activity context, though, as it may not be obvious with future app changes.

**Comment by BumbleFlash** [comment]
@barbeau from a brief reading about Dagger, in order to inject a context variable, the class must extend `DaggerAppComponent` to bind the context. Since `MainActivity` already extends `BaseActivity`, I'm not quite sure how I can extend Dagger classes.

**Comment by barbeau** [comment]
@BumbleFlash Shouldn't the injection be happening for `PlaylistManager`?

**Comment by BumbleFlash** [comment]
Per discussion, we think injecting `context` with dagger could be more complicated than we thought hence we have decided to manually pass in the `context` in the method signature to fix the crash. (Implemented in #107)
